pkg/queue/sharedmain/main.go

```diff
@@ -243,52 +243,53 @@ func Main(opts ...Option) error {
 		concurrencyendpoint = queue.NewConcurrencyEndpoint(env.ConcurrencyStateEndpoint, env.ConcurrencyStateTokenPath)
 	}
 
+	type service struct {
+		name string
+		srv  *http.Server
+		tls  bool
+	}
+
+	var srvs []service
+	var drainer *pkghandler.Drainer
+	var mainServer *http.Server
+
 	// Enable TLS when certificate is mounted.
 	tlsEnabled := exists(logger, certPath) && exists(logger, keyPath)
 
-	mainServer, drainer := buildServer(d.Ctx, env, d.Transport, probe, stats, logger, concurrencyendpoint, false)
-	httpServers := map[string]*http.Server{
-		"main":    mainServer,
-		"metrics": buildMetricsServer(protoStatReporter),
-		"admin":   buildAdminServer(d.Ctx, logger, drainer),
-	}
-	if env.EnableProfiling {
-		httpServers["profile"] = profiling.NewServer(profiling.NewHandler(logger, true))
+	if tlsEnabled {
+		mainServer, drainer = buildServer(d.Ctx, env, d.Transport, probe, stats, logger, concurrencyendpoint, true)
+		srvs = append(srvs, service{name: "tlsMain", srv: mainServer, tls: true})
+		srvs = append(srvs, service{name: "tlsAdmin", srv: buildAdminServer(d.Ctx, logger, drainer), tls: true})
+	} else {
+		mainServer, drainer = buildServer(d.Ctx, env, d.Transport, probe, stats, logger, concurrencyendpoint, false)
+		srvs = append(srvs, service{name: "main", srv: mainServer, tls: false})
+		srvs = append(srvs, service{name: "admin", srv: buildAdminServer(d.Ctx, logger, drainer), tls: false})
 	}
 
-	// Enable TLS server when activator server certs are mounted.
-	// At this moment activator with TLS does not disable HTTP.
-	// See also https://github.com/knative/serving/issues/12808.
-	var tlsServers map[string]*http.Server
-	if tlsEnabled {
-		mainTLSServer, drainer := buildServer(d.Ctx, env, d.Transport, probe, stats, logger, concurrencyendpoint, true /* enable TLS */)
-		tlsServers = map[string]*http.Server{
-			"tlsMain":  mainTLSServer,
-			"tlsAdmin": buildAdminServer(d.Ctx, logger, drainer),
-		}
-		// Drop admin http server as we Use TLS for the admin server.
-		// TODO: The drain created with mainServer above is lost. Unify the two drain.
-		delete(httpServers, "admin")
+	srvs = append(srvs, service{name: "metrics", srv: buildMetricsServer(protoStatReporter), tls: false})
+
+	if env.EnableProfiling {
+		srvs = append(srvs, service{name: "profile", srv: profiling.NewServer(profiling.NewHandler(logger, true)), tls: false})
 	}
 
 	logger.Info("Starting queue-proxy")
 
 	errCh := make(chan error)
-	for name, server := range httpServers {
-		go func(name string, s *http.Server) {
+	for _, s := range srvs {
+		go func(s service) {
+			logger.Debugf("Starting service %s on port %s with tls %t", s.name, s.srv.Addr, s.tls)
 			// Don't forward ErrServerClosed as that indicates we're already shutting down.
-			if err := s.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
-				errCh <- fmt.Errorf("%s server failed to serve: %w", name, err)
+			var err error
+			if s.tls {
```

Comment:
You could use some naming convention to make this more compact:

```
if strings.HasPrefix(s.name, "tls") {
...
 } else {
...
}
```
Another way would be funcs like buildAdminServer, buildServer to return the `service` construct instead.

Reply:
ok, lets have buildAdminServer, buildServer, buildMetricsServer return service, and add buildProfile that will also add service. 

I would prefer a flag on a naming prefix for TLS to make it explicit and avoid future bugs.